### DISCUSSION TITLE
Strange behavior of JSON field and `commit()`

### DISCUSSION THREAD

**Discussion by matmarczak** [sqlalchemy/sqlalchemy#7886] [Usage Questions] [answered]
Hi!

I recently found strange behavior and count on you to explain it. I'm not able to provide reproducible example because couldn't extract it. Its strange one related to `dict` mutation in `JSON` field.

Lets say there exists table that has JSON field named `jfield`. Usually when I modify `jfield` (by dict assignment, update in place etc.) and do `commit()`, changes are written to database. But I found one case, when modified `jfield` is not committed to database after calling `commit()` (no exception raised either). In this case I can see that `session.dirty` contains changed instance with changed `jfield`, but after `commit()` nothing is commited to DB. I can workaround this by `deepcopy`ing entire `jfield` and assigning to `instance.jfield` again. When I assign `deepcopy` version and do `commit()` the same change is written to database. Worth to mention, that `jfield` changes its value in meaningful way (`jfield != jfield'`).

There might be some issue with my usage of `JSON` field, but also want to make sure its not something in SQLAlchemy that others will bump into.

**Comment by zzzeek** [comment, marked as answer]
in-place mutations are not tracked by the ORM automatically.  you probably should take a look at https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html

**Reply by CaselIT** [reply]
As an alternative, in case of a dict it should be sufficient to do a shallow of the content for the orm to properly track the change, without needing a full deep copy.

Another alternative is to manually mark the attribute as modified with https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=dirty#sqlalchemy.orm.attributes.flag_modified
